In [1]:
import pandas as pd
import tarfile
import csv

tar = tarfile.open("/Users/gavincleach/Desktop/famous-chord-progressions-raw-data/billboard-2.0-salami_chords.tar.gz", "r:gz")
famous_chord_progressions_coded = []
for member in tar.getmembers():
    f = tar.extractfile(member)
    famous_chord_progressions_coded.append(f.read())


famous_chord_progressions = []

for song in famous_chord_progressions_coded:
    famous_chord_progressions.append(song.decode("utf-8"))
    
    

billboard_charts = pd.read_csv('/Users/gavincleach/Desktop/famous-chord-progressions-raw-data/billboard-2.0-index.csv')
        
len(famous_chord_progressions)

890

In [2]:
billboard_charts.set_index("id", inplace = True)

billboard_charts.head()

chart_date  target_rank  actual_rank                title  \
id                                                              
1   1987-07-11           82          NaN                  NaN   
2   1973-08-04           14          NaN                  NaN   
3   1961-07-03           56         57.0         I Don't Mind   
4   1971-08-07           32         31.0  You've Got A Friend   
5   1976-07-10           51          NaN                  NaN   

                          artist  peak_rank  weeks_on_chart  
id                                                           
1                            NaN        NaN             NaN  
2                            NaN        NaN             NaN  
3                    James Brown       47.0             8.0  
4   Roberta Flack,Donny Hathaway       29.0            12.0  
5                            NaN        NaN             NaN

let's first get rid of the rows in `billboard_charts` that have `NaN` value in the `title` column. There are 890 songs with full chord progressions, so we will see how many entries we have in `billboard_charts` after we drop those rows.

In [3]:
unwanted_portion = billboard_charts.loc[billboard_charts['title'].isnull()]
ids_to_drop = list(unwanted_portion.index) 

billboard_charts = billboard_charts.drop(labels = ids_to_drop, axis = 0)

billboard_charts.shape

(890, 7)

Great! So now we only have the remaining song entries that don't have `NaN` entries.

Before we clean up `billboard_charts`, we should clean up the `famous_chord_progressions` list, because there is much more work to do on that.

It would be beneficial to understand what each item in the list `famous_chord_progressions` means. So...

(ignore \'s)

\# title: \
after this comes the title of the song. 

\# artist: \
after this comes the artist name.

\# metre: \
after this comes the time signature.

\# tonic: \
after this comes the key the song starts in. 

for tonic and metre there my not be a definite answer for each of these, in that case, there will be a '?' in place. Tonic and metre might change throughout the song as well.

In [7]:
import re

song_example = famous_chord_progressions[0]
part_of_song_example = "# title: I Don't mind\n# artist: Jam"

test_str_0 = 'my favorite numbers are 16 and 2\n'
#test_str_1 = 

#
title_pattern = "# title: (.+)\n"

artist_part_of_song_example = song_example.split(' artist: ')[1]

artist_pattern = "(.+)\n"

metre_part_of_song_example = song_example.split(' metre: ')[1]

metre_pattern = "(.+)\n"

tonic_part_of_song_example = song_example.split(' tonic: ')[1]

tonic_pattern = "([ABCDEFG]b?)\n"

comma_split_list = re.split(', ?', song_example)[1:]
refined_comma_split_list = []

for portion in comma_split_list:
    if re.match("\([a-z]+", portion):
        refined_comma_split_list.append(refined_comma_split_list[-1] + ' ' + portion)
        refined_comma_split_list.pop(-2)
    elif re.match("[a-z]+\)", portion):
        refined_comma_split_list.append(refined_comma_split_list[-1] + ' ' + portion)
        refined_comma_split_list.pop(-2)
    else:
        refined_comma_split_list.append(portion)
print(song_example)
print('\n\n\n')
print(refined_comma_split_list)  
print('\n\n\n')
    
        

first_part_result = refined_comma_split_list[0]

first_part_of_song_chords = refined_comma_split_list[1]

second_part_result = refined_comma_split_list[2]

second_part_of_song_chords = refined_comma_split_list[3]

third_part_result = refined_comma_split_list[4]

third_part_of_song_chords = refined_comma_split_list[5]

fourth_part_result = refined_comma_split_list[6]

fourth_part_of_song_chords = refined_comma_split_list[7]

fifth_part_result = refined_comma_split_list[8]

fifth_part_of_song_chords = refined_comma_split_list[9]

sixth_part_result = refined_comma_split_list[10]

sixth_part_of_song_chords = refined_comma_split_list[11]

chords_pattern = "([ABCDEFG][#b]?\S+) \|?"

#maybe the part will repeat itself!
repetition_pattern = "x(\d+)"






title_result = re.findall(title_pattern, part_of_song_example)[0]
artist_result = re.findall(artist_pattern, artist_part_of_song_example)[0]
metre_result = re.findall(metre_pattern, metre_part_of_song_example)[0]
tonic_result = re.findall(tonic_pattern, tonic_part_of_song_example)[0]

#first_part_result
first_part_chords_result = re.findall(chords_pattern, first_part_of_song_chords)

if len(re.findall(repetition_pattern, first_part_of_song_chords)) == 0:
    first_part_repeat_result = 1
else:
    first_part_repeat_result = re.findall(repetition_pattern, first_part_of_song_chords)[0]

#second_part_result
second_part_chords_result = re.findall(chords_pattern, second_part_of_song_chords)

if len(re.findall(repetition_pattern, second_part_of_song_chords)) == 0:
    second_part_repeat_result = 1
else:
    second_part_repeat_result = re.findall(repetition_pattern, second_part_of_song_chords)[0]

#third_part_result
third_part_chords_result = re.findall(chords_pattern, third_part_of_song_chords)

if len(re.findall(repetition_pattern, third_part_of_song_chords)) == 0:
    third_part_repeat_result = 1
else:
    third_part_repeat_result = re.findall(repetition_pattern, third_part_of_song_chords)[0]

#fourth_part_result
fourth_part_chords_result = re.findall(chords_pattern, fourth_part_of_song_chords)

if len(re.findall(repetition_pattern, fourth_part_of_song_chords)) == 0:
    fourth_part_repeat_result = 1
else:
    fourth_part_repeat_result = re.findall(repetition_pattern, fourth_part_of_song_chords)[0]

#fifth_part_result
fifth_part_chords_result = re.findall(chords_pattern, fifth_part_of_song_chords)

if len(re.findall(repetition_pattern, fifth_part_of_song_chords)) == 0:
    fifth_part_repeat_result = 1
else:
    fifth_part_repeat_result = re.findall(repetition_pattern, fifth_part_of_song_chords)[0]

#sixth_part_result
sixth_part_chords_result = re.findall(chords_pattern, sixth_part_of_song_chords)

if len(re.findall(repetition_pattern, sixth_part_of_song_chords)) == 0:
    sixth_part_repeat_result = 1
else:
    sixth_part_repeat_result = re.findall(repetition_pattern, sixth_part_of_song_chords)[0]


print(title_result)
print(artist_result)
print(metre_result)
print(tonic_result)

print(first_part_result)
print(first_part_chords_result)
print(first_part_repeat_result)

print(second_part_result)
print(second_part_chords_result)
print(second_part_repeat_result)

print(third_part_result)
print(third_part_chords_result)
print(third_part_repeat_result)

print(fourth_part_result)
print(fourth_part_chords_result)
print(fourth_part_repeat_result)

print(fifth_part_result)
print(fifth_part_chords_result)
print(fifth_part_repeat_result)

print(sixth_part_result)
print(sixth_part_chords_result)
print(sixth_part_repeat_result)

I Don't mind
James Brown
6/8
C


["# title: I Don't mind\n# artist: James Brown\n#",
 '6/8\n# tonic: C\n\n0.0\tsilence\n7.3469387e-2\tA, intro, | A:min | A:min | C:maj | C:maj |\n8.714013605\t| A:min | A:min | C:maj | C:maj |\n15.611995464\t| A:min | A:min | C:maj | C:maj |\n22.346394557\tB, verse, | A:min | A:min | C:maj | C:maj |, (voice\n29.219433106\t| A:min | A:min | C:maj | C:maj |\n36.279501133\t| F:maj | F:maj | D:maj | D:maj |\n43.053922902\t| G:maj | G:maj | C:maj | C:maj |\n49.23802721\tB, verse, | A:maj | A:maj | C:maj | C:maj |\n56.185079365\t| A:maj | A:maj | C:maj | C:maj |\n63.007732426\t| F:maj | F:maj | D:maj | D:maj |\n69.942857142\t| G:maj | G:maj | C:maj | C:maj |, voice)\n76.123990929\tA, interlude, | A:min | A:min | C:maj | C:maj |\n82.811201814\t| A:min | A:min | C:maj | C:maj |\n89.535532879\t| F:maj | F:maj | D:maj | D:maj |\n96.418956916\t| G:maj | G:maj | C:maj | C:maj |, (guitar)\n102.924353741\tB, verse, | A:maj | A:maj | C:maj | C:maj |, (voice\n109.886984126\t| A:maj | A:maj | C:maj | C

In [6]:
famous_chord_progressions[0:90]

["# title: I Don't mind\n# artist: James Brown\n# metre: 6/8\n# tonic: C\n\n0.0\tsilence\n7.3469387e-2\tA, intro, | A:min | A:min | C:maj | C:maj |\n8.714013605\t| A:min | A:min | C:maj | C:maj |\n15.611995464\t| A:min | A:min | C:maj | C:maj |\n22.346394557\tB, verse, | A:min | A:min | C:maj | C:maj |, (voice\n29.219433106\t| A:min | A:min | C:maj | C:maj |\n36.279501133\t| F:maj | F:maj | D:maj | D:maj |\n43.053922902\t| G:maj | G:maj | C:maj | C:maj |\n49.23802721\tB, verse, | A:maj | A:maj | C:maj | C:maj |\n56.185079365\t| A:maj | A:maj | C:maj | C:maj |\n63.007732426\t| F:maj | F:maj | D:maj | D:maj |\n69.942857142\t| G:maj | G:maj | C:maj | C:maj |, voice)\n76.123990929\tA, interlude, | A:min | A:min | C:maj | C:maj |\n82.811201814\t| A:min | A:min | C:maj | C:maj |\n89.535532879\t| F:maj | F:maj | D:maj | D:maj |\n96.418956916\t| G:maj | G:maj | C:maj | C:maj |, (guitar)\n102.924353741\tB, verse, | A:maj | A:maj | C:maj | C:maj |, (voice\n109.886984126\t| A:maj | A:maj | C:maj 